File: MainProject/Scenes/HexPlanet/Utils/TilePriorityQueue.cs
using System.Collections.Generic;
using ZeromaXsPlaygroundProject.Scenes.HexPlanet.Structs;

namespace ZeromaXsPlaygroundProject.Scenes.HexPlanet.Utils;

/// Copyright (C) 2025 Zhu Xiaohe(aka ZeromaXHe)
/// Author: Zhu XH
/// Date: 2025-03-03 09:14
public class TilePriorityQueue(TileSearchData[] data)
{
    private readonly List<int> _list = [];
    private int _minimum = int.MaxValue;

    public void Enqueue(int tileId)
    {
        var priority = data[tileId].SearchPriority;
        if (priority < _minimum)
            _minimum = priority;
        while (priority >= _list.Count)
            _list.Add(-1);
        data[tileId].NextWithSamePriority = _list[priority];
        _list[priority] = tileId;
    }
    
    public bool TryDequeue(out int tileId)
    {
        for (; _minimum < _list.Count; _minimum++)
        {
            tileId = _list[_minimum];
            if (tileId >= 0)
            {
                _list[_minimum] = data[tileId].NextWithSamePriority;
                return true;
            }
        }
        tileId = -1;
        return false;
    }

    public void Change(int tileId, int oldPriority)
    {
        var current = _list[oldPriority];
        var next = data[current].NextWithSamePriority;
        if (current == tileId)
            _list[oldPriority] = next;
        else
        {
            while (next != tileId)
            {
                current = next;
                next = data[current].NextWithSamePriority;
            }
            data[current].NextWithSamePriority = data[tileId].NextWithSamePriority;
        }
        Enqueue(tileId);
    }
    
    public void Clear()
    {
        _list.Clear();
        _minimum = int.MaxValue;
    }
}

File: TerraObserver/TO.Repos/Data/PathFindings/TilePriorityQueue.fs
namespace TO.Repos.Data.PathFindings

open System
open TO.Domains.Alias.HexSpheres.Tiles
open TO.Domains.Structs.PathFindings

/// Copyright (C) 2025 Zhu Xiaohe(aka ZeromaXHe)
/// Author: Zhu XH (ZeromaXHe)
/// Date: 2025-06-06 21:19:06
type TilePriorityQueue(data: TileSearchData array) =
    let list = ResizeArray<int>()
    let mutable minimum = Int32.MaxValue

    member this.Enqueue(tileId: TileId) =
        let priority = data[tileId].SearchPriority

        if priority < minimum then
            minimum <- priority

        while priority >= list.Count do
            list.Add -1

        data[tileId].NextWithSamePriority <- list[priority]
        list[priority] <- tileId

    member this.TryDequeue(tileId: TileId outref) =
        let mutable result = false

        while minimum < list.Count && not result do
            tileId <- list[minimum]

            if tileId >= 0 then
                list[minimum] <- data[tileId].NextWithSamePriority
                result <- true
            else
                minimum <- minimum + 1

        if not result then
            tileId <- -1

        result

    member this.Change(tileId: TileId, oldPriority: int) =
        let mutable current = list[oldPriority]
        let mutable next = data[current].NextWithSamePriority

        if current = tileId then
            list[oldPriority] <- next
        else
            while next <> tileId do
                current <- next
                next <- data[current].NextWithSamePriority

            data[current].NextWithSamePriority <- data[tileId].NextWithSamePriority

        this.Enqueue tileId

    member this.Clear() =
        list.Clear()
        minimum <- Int32.MaxValue
